PubSub/src/BatchPublisher.php

```diff
@@ -98,11 +109,54 @@ public function publish(array $message)
      */
     protected function getCallback()
     {
+        return [$this, $this->batchMethod];
+    }
+
+    /**
+     * Publish a set of deferred messages, sorted into multiple calls by ordering key.
+     *
+     * Intended for internal use only by the batch publisher.
+     *
+     * @see https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.topics/publish Publish Message
+     *
+     * @param array[] $messages A list of messages. Each message must be in the correct
+     *        [Message Format](https://cloud.google.com/pubsub/docs/reference/rest/v1/PubsubMessage).
+     * @param array $options [optional] Configuration Options
+     * @return array A list of message IDs.
+     * @internal
+     * @access private
+     */
+    public function publishDeferred(array $messages, array $options = [])
```

Comment:
How are we ensuring that there is only a single message batch outstanding to the server at a time? Is it only possible to have one call to publishDeferred outstanding at a time?

Reply:
The batch daemon uses shared memory and semaphores to control concurrency. There will only be a single batch active on a single server at a given time.